flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ComparableTypeStrategy.java

```diff
@@ -0,0 +1,225 @@
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.hasFamily;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.hasRoot;
+
+/**
+ * An {@link InputTypeStrategy} that checks if all input arguments can be compared with each other with
+ * the minimal provided comparision.
+ *
+ * <p>It requires at least two arguments.
+ *
+ * <p>For the rules which types are comparable with which types see {@link #areComparable(LogicalType, LogicalType)}.
+ */
+@Internal
+public final class ComparableTypeStrategy implements InputTypeStrategy {
+	private final StructuredComparision requiredComparision;
+	private final ArgumentCount argumentCount;
+
+	public ComparableTypeStrategy(
+			ArgumentCount argumentCount,
+			StructuredComparision requiredComparision) {
+		Preconditions.checkArgument(
+			argumentCount.getMinCount().map(c -> c >= 2).orElse(false),
+			"Comparable type strategy requires no less than two arguments. Actual minimal argument count: %s",
+			argumentCount.getMinCount().map(Objects::toString).orElse("<None>"));
+		Preconditions.checkArgument(requiredComparision != StructuredComparision.NONE);
+		this.requiredComparision = requiredComparision;
+		this.argumentCount = argumentCount;
+	}
+
+	@Override
+	public ArgumentCount getArgumentCount() {
+		return argumentCount;
+	}
+
+	@Override
+	public Optional<List<DataType>> inferInputTypes(CallContext callContext, boolean throwOnFailure) {
+		List<DataType> argumentDataTypes = callContext.getArgumentDataTypes();
+		for (int i = 0; i < argumentDataTypes.size() - 1; i++) {
+			LogicalType firstLogicalType = argumentDataTypes.get(i).getLogicalType();
+			LogicalType secondLogicalType = argumentDataTypes.get(i + 1).getLogicalType();
+
+			if (!areComparable(firstLogicalType, secondLogicalType)) {
+				if (throwOnFailure) {
+					throw callContext.newValidationError(
+						"All types in EQUALS should support %s comparison with each other. Can not compare %s with %s",
+						requiredComparision == StructuredComparision.EQUALS ? "'EQUALS'" : "both 'EQUALS' and 'ORDER'",
+						firstLogicalType,
+						secondLogicalType
+					);
+				}
+
+				return Optional.empty();
+			}
+		}
+
+		return Optional.of(argumentDataTypes);
+	}
+
+	private boolean areComparable(LogicalType firstLogicalType, LogicalType secondLogicalType) {
+		// A hack to support legacy types. To be removed when we drop the legacy types.
+		if (firstLogicalType instanceof LegacyTypeInformationType ||
+				secondLogicalType instanceof LegacyTypeInformationType) {
+			return true;
+		}
+
+		// everything is comparable with null, it should return null in that case
+		if (hasRoot(firstLogicalType, LogicalTypeRoot.NULL) || hasRoot(secondLogicalType, LogicalTypeRoot.NULL)) {
+			return true;
+		}
+
+		if (hasRoot(firstLogicalType, LogicalTypeRoot.STRUCTURED_TYPE) &&
```

Comment:
nit: check for same type root and then use a switch/case in a separate method?

Reply:
Very good comment! It simplified the code a lot!